### PR TITLE
WIP - Search lists by category

### PR DESCRIPTION
**Pull request by wsf246**
**Problem:**
No way to filter lists#index for just categories you want to find.

**Solution:**
Eff with Vue til it filters stuff

**Additional Notes:**
Sorry for the giant commit, I had to do a lot of playing around to properly understand how to make Vue work.  I couldn't figure out how to get vue to work with ERB renders, I couldn't properly pass around the vue objects into the partials and vice versa so I had to put list back into list#index.  

I think this could be solved by using vue components?  A lot of the issues I've been having is figuring out how to integrate view with Rails/ERB and the asset pipeline.  I think we need to add some gems or other things to get it to properly load?  

I was also playing around with grabbing the list#index json data with ajax.  I think that will be the long term solution.  We can also separate pinned and unpinned lists through filter or computed at that point as well.  

This is obviously a WIP at this point.  With this v-model setup though, we can also add an `<input>` so you just type in a tag for lists.  I think we should have a category drop down and a type in field for now with eventual integration with whatever search gem we end up using.

### CODE REVIEW

**Comment by briankung** [comment]
We should probably be doing a lot more work in the Vue script and less in the Erb.

**Comment by briankung** [comment]
I would probably have spit out the list data in JSON in the Vue script and had the Vue script populate the template.

Something like

```js
new Vue({
  data: {
    lists: [
      {
        # ... list data
      }
    ]
  }
});
```

And generate the Vue components from there.

**Comment by briankung** [comment]
And in Erb:

```js
new Vue({
  data: {
    lists: <%= @lists.to_json %>
  }
});
```

**Comment by wsf246** [comment]
Yea the issue w that is i lose a lot of rails functionality, like i need
list.tag_list

**Comment by wsf246** [comment]
I can add it in the controller though, so maybe ill try that.

File: app/views/lists/index.html.erb [hunk 1]
```diff
@@ -4,36 +4,125 @@
...
+      <div class="col-md-3">
+        <%= link_to new_list_path, class: "btn panel panel-default new-list" do %>
+          <h2>Create a new list</h2>
+        <% end %>
+      </div>
+
+      <%# take out render to have vue app on one page %>
+      <% @pinned_lists.each do |list|%>
+        <%# list_partial %>
+        <div class="col-md-3 list-card" v-show="<%= list.tag_list.to_a %>.includes(category.toLowerCase()) || category == 'All Categories'">
```
Comment (briankung) [inline app/views/lists/index.html.erb]: Whoa, this is crazy. I would much prefer this be in a "search" method or something on the Vue object so we know where it is.

File: app/views/lists/index.html.erb [hunk 2]
```diff
@@ -4,36 +4,125 @@
...
+                      </small>
+                    <% end %>
+                  </div>
+                </div>
+              <% end %>
+              <h2>
+                <%= link_to list.name, user_list_path(list.owner, list) %><br>
+                <small> by <%= link_to list.owner.username, user_lists_path(list.owner) %></small>
+              </h2>
+              <p><%= truncate(list.description, length: 400, separator: ' ') %></p>
```
Comment (briankung) [inline app/views/lists/index.html.erb]: A lot of this logic can be put into Vue now

File: app/views/lists/index.html.erb (hunk 1 above)
Comment (wsf246) [inline app/views/lists/index.html.erb]: yes, I agree, I think that's what i'll try to do, my issue is accessing the list for the @pinned_lists.each.  I think this is fixed once we put all the lists in an array like you said above and use vue to make it work.  maybe let's jump on a screenhero and see if we're on the same page on how to do this cuz im still getting used to vue.

File: app/views/lists/index.html.erb [hunk 3]
```diff
@@ -4,36 +4,77 @@
...
-        <h2>Create a new list</h2>
-      <% end %>
+    <div class="cards row">
+      <div class="col-md-3">
+        <%= link_to new_list_path, class: "btn panel panel-default new-list" do %>
+          <h2>Create a new list</h2>
+        <% end %>
+      </div>
+
+      <%# take out render to have vue app on one page %>
```
Comment (briankung) [inline app/views/lists/index.html.erb]: I don't think you have to pull the lists partial into this parent template, either. You can have the Vue app on this page and have the `lists/list` Erb template have the Vue directives.

Comment (briankung) [inline app/views/lists/index.html.erb]: I'm pretty sure we do not want to take out the render.

**Comment by wsf246** [comment]
I'll fix all this, I'm just playing around to clean it up right now.

**Comment by briankung** [comment]
There was a [problem deploying](https://dashboard.heroku.com/pipelines/a9cc69d8-ada6-4fa3-a871-58c7c340eeaa/review-app/2ac341e7-f279-4dc6-a99e-fbf615dcbe05) the review app.

Looks like this is the culprit:

> /app/app/models/user.rb:89:in `subscribe_user_to_all_users_list'

**Comment by briankung** [comment]
> Gibbon::GibbonError: You must set an api_key prior to making a call

So, to set environment variables for review apps, you need to add it to `app.json` similar to the google analytics id:

```json
    "GA_TRACKING_ID": {
      "required": true
    }
```

That forces the review apps to inherit the environment variable from staging. Speaking of which, you will need to set the variable in staging. I would set it to something meaningless so that the API calls fail. That's fine. We don't want to actually send staging data to Mailchimp.

File: app/views/lists/index.html.erb [hunk 4]
```diff
@@ -4,36 +4,67 @@
...
+          [
+            'All Categories',
+            'Health and Nutrition',
+            'Economics and Social Policy',
+            'Science and Technology',
+            'Other'
+          ]
+      },
+      methods: {
+        filterByTag: function(tags) {
```
Comment (briankung) [inline app/views/lists/index.html.erb]: This should probably be named something like `showSelectedCategory`. It's not really a filter in the sense of a Vue filter, and it's also not a filter in the sense that it returns a boolean, instead of reducing the original input according to some criteria.

File: app/views/lists/_list.html.erb [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<div class="col-md-3 list-card">
+<div class="col-md-3 list-card" v-show="filterByTag(<%= list.tag_list.to_a %>)">
```
Comment (briankung) [inline app/views/lists/_list.html.erb]: I'm very curious about how this renders. In order for it to work, `filterByTag(<%= some_array.to_a %>)` needs to output `filterByTag(["item1", "item2])`.

Does this work? It's kind of coincidental that Ruby and Javascript arrays share the same syntax, which makes me a bit nervous. If it works, I suppose it's fine and I could get used to it, but it does seem like programming by coincidence.

**Comment by wsf246** [comment]
programming by coincidence is my favorite way to program.

File: app/views/lists/_list.html.erb (hunk 5 above)
Comment (wsf246) [inline app/views/lists/_list.html.erb]: I tried just send `<%= list.tag_list %>` into the method, vue was reading each of the items as a vue variable.  This does work though.

File: app/views/lists/index.html.erb (hunk 4 above)
Comment (wsf246) [inline app/views/lists/index.html.erb]: I also have a development list and mailchimp API key

File: app/views/lists/index.html.erb [hunk 6]
```diff
@@ -4,36 +4,79 @@
...
+          if (this.selectedCategory === 'All Categories') {
+            return true
+          } else if (this.selectedCategory === 'Untagged') {
+            for (var i = 0; i < this.categories.length; i++) {
+              if (tags.includes(this.categories[i].toLowerCase())) {
+                return false
+                break
+              }
+              return true
+            }
```
Comment (wsf246) [inline app/views/lists/index.html.erb]: for the life of my I can't figure this out.  I'm trying to create an untagged category that shows all list that don't have the tags in the category (I'll remove 'All Categories' and 'Untagged' after I figure this out).  

But this for loop just won't work.  The odd thing is this.categories[1] or this.categories[2] works, but if I use the `i` variable, it doesn't do it.  Any idea what's going on?

Comment (wsf246) [inline app/views/lists/index.html.erb]: I tested the for loop logic in console and it works, so I'm super confused.

Comment (wsf246) [inline app/views/lists/index.html.erb]: also tried `Number(i)` which did nothing.

File: db/seeds.rb [hunk 7]
```diff
@@ -37,7 +37,7 @@
 
 
 ActiveRecord::Base.transaction do
-  u = User.create(email: 'user@example.com', password: 'password', username: 'testuser')
+  u = User.create(email: 'testaccount@projectcredo.com', password: 'password', username: 'testuser')
```
Comment (wsf246) [inline db/seeds.rb]: fyi, mailchimp rejects fake looking emails.

Comment (wsf246) [inline db/seeds.rb]: now that I think of it, this may only work once once testaccount@projectcredo.com is signed up

Comment (briankung) [inline db/seeds.rb]: Should probably devise a way to disable subscribing users to the list. Especially test users.

File: app/views/lists/index.html.erb (hunk 6 above)
Comment (briankung) [inline app/views/lists/index.html.erb]: I'm going to debug, but start with logging everything to console - `i`, any variables, `this`, `this.categories`

**Comment by wsf246** [comment]
ah good point.

**Comment by briankung** [comment]
- Category selector looks like it should be an input field
- This seems more like a filter to me, and I think filters are generally expected to be on the left hand side in a sidebar (a la Amazon, Newegg, other shopping sites)
- The problem with filtering not working with those specific categories could be sidestepped by using an input field with autocomplete on a list of available existing tags.

File: app/views/lists/index.html.erb [hunk 8]
```diff
@@ -4,36 +4,108 @@
...
+      data: {
+        selectedTag: 'All Tags',
+        query: '',
+        tags: <%=  @tag_buckets  %>,
+        results: [],
+        placeholder: "Search for a list"
+      },
+      computed: {
+        matchingTags() {
+          return this.tags.filter(tag => tag.toLowerCase().includes(this.query.toLowerCase())).slice(0,10)
```
Comment (briankung) [inline app/views/lists/index.html.erb]: Would you mind breaking this into separate lines for readability?

File: app/views/lists/index.html.erb [hunk 9]
```diff
@@ -4,36 +4,108 @@
...
+          return listTags.includes(this.selectedTag.toLowerCase()) || this.selectedTag === 'All Tags'
+        },
+        showQueriedTag: function(tags) {
+          return tags.includes(this.query.toLowerCase()) || this.query === ''
+        },
+        getResults: function() {
+          if (this.query === '') {
+            return this.results = []
+          } else {
+            return this.results = this.matchingTags
```
Comment (briankung) [inline app/views/lists/index.html.erb]: I don't think you need to return here. I think I was doing that in the crossref `getResults` method because I wanted to be sure execution was ending there.

File: app/controllers/lists_controller.rb [hunk 10]
```diff
@@ -4,6 +4,9 @@ class ListsController < ApplicationController
   # GET /lists
   # GET /lists.json
   def index
+    all_generated_tags = List.publicly_visible.map(&:tag_list).flatten.uniq
+    titleized_tags = all_generated_tags.map!(&:titleize).sort!
+    @tag_buckets = titleized_tags.to_json.html_safe
```
Comment (briankung) [inline app/controllers/lists_controller.rb]: Can we just do `@tags = Tag.all` for simplicity's sake? And then format them to json and whatnot in the view.

Why are you doing `html_safe`?

File: app/views/lists/index.html.erb [hunk 11]
```diff
@@ -4,36 +4,108 @@
...
+<!--   <div class="form-horizontal">
+    <div class="form-group">
+      <label class="col-md-1 control-label">Category:</label>
+      <div class="col-md-2">
+        <select class="form-control" v-model="selectedTag">
+          <option v-for="tag in tags" v-text="tag"></option>
+        </select>
+      </div>
+    </div>
+  </div> -->
```
Comment (briankung) [inline app/views/lists/index.html.erb]: This will be cleaned up, I presume?

File: app/views/lists/index.html.erb [hunk 12]
```diff
@@ -4,36 +4,108 @@
...
+        results: [],
+        placeholder: "Search for a list"
+      },
+      computed: {
+        matchingTags() {
+          return this.tags.filter(tag => tag.toLowerCase().includes(this.query.toLowerCase())).slice(0,10)
+        }
+      },
+      methods: {
+        showSelectedTag: function(listTags) {
```
Comment (briankung) [inline app/views/lists/index.html.erb]: Will you use this method later? If not, make sure to remove it.

File: app/views/lists/index.html.erb [hunk 13]
```diff
@@ -4,36 +4,108 @@
...
+          return tags.includes(this.query.toLowerCase()) || this.query === ''
+        },
+        getResults: function() {
+          if (this.query === '') {
+            return this.results = []
+          } else {
+            return this.results = this.matchingTags
+          }
+        },
+        selectResult: function(result) {
```
Comment (briankung) [inline app/views/lists/index.html.erb]: We don't need to `selectResult` since we're not submitting anything, unlike the crossref example.

File: db/seeds.rb (hunk 7 above)
Comment (briankung) [inline db/seeds.rb]: I also think we should revert this change if we're going to prevent subscription for the test user. I don't want to even accidentally add to our spam load and `example.com` is specifically for testing.

File: app/views/lists/index.html.erb (hunk 9 above)
Comment (briankung) [inline app/views/lists/index.html.erb]: Wait, no I wasn't, because the final version of the Crossref `getResults` doesn't do that. At some point, I was doing that during testing, though.

Methods mutate variables. Computed properties are the only ones that require a return statement. Not sure about watchers, actually, but I don't think they need to return anything, either.

File: app/controllers/lists_controller.rb (hunk 10 above)
Comment (wsf246) [inline app/controllers/lists_controller.rb]: I don't think I can access Tag.all and that would also grab all the tags we have on papers too, which could get confusing until we start getting more complicated on how we search and what we want to return. I'm doing html_safe because the original array was do doing &quot instead of a double quote.

I can move the formatting to vue to handle.  Probably format most of it with js?  There are a lot of nifty formatting stuff in rails though like titleize which I don't know if there's a function like that in js.

File: app/views/lists/index.html.erb (hunk 8 above)
Comment (wsf246) [inline app/views/lists/index.html.erb]: sure, this is the ES6 format? or something, I got it from the Vue laracast.  I can use the old formatting as well if it's more readable.

File: app/views/lists/index.html.erb (hunk 9 above)
Comment (wsf246) [inline app/views/lists/index.html.erb]: so get rid of the returns and keep this in method?

File: app/views/lists/index.html.erb (hunk 11 above)
Comment (wsf246) [inline app/views/lists/index.html.erb]: yea, if this autocomplete stuff is good, then I was going to get rid of all the select stuff.  unless we wanted to do both,

File: app/views/lists/index.html.erb (hunk 12 above)
Comment (wsf246) [inline app/views/lists/index.html.erb]: probably get rid of it if we just want to do the autocomplete select,

File: app/views/lists/index.html.erb (hunk 13 above)
Comment (wsf246) [inline app/views/lists/index.html.erb]: this is to select the result from the list so you don't have to type everything out.

File: app/controllers/lists_controller.rb (hunk 10 above)
Comment (briankung) [inline app/controllers/lists_controller.rb]: I think you can do `ActsAsTaggable::Tag.all`. Or something more complicated, since they're joined by `Taggings`. Although showing only tags for visible lists _is_ nice.

Also, if the user is logged in, you probably don't want just publicly visible lists, right? Maybe check if they're logged in and use the `visible_to` scope on List.

Try this instead: `@tags = all_generated_tags.map(&:titleize).sort` and then use `to_json` in the view. I somehow don't think you need `html_safe` in the view. Also, this is bad if someone inputs an html element as a tag.

File: app/views/lists/index.html.erb (hunk 13 above)
Comment (briankung) [inline app/views/lists/index.html.erb]: Oh...good point. Dang, that's kinda weird. You can't hit enter to complete the phrase. Oh well.

File: app/views/lists/index.html.erb (hunk 9 above)
Comment (briankung) [inline app/views/lists/index.html.erb]: Yes

File: app/views/lists/index.html.erb (hunk 8 above)
Comment (briankung) [inline app/views/lists/index.html.erb]: This is ES6, which we should probably get used to, but that's independent of breaking into separate lines.

File: app/views/lists/index.html.erb (hunk 11 above)
Comment (briankung) [inline app/views/lists/index.html.erb]: I wouldn't call the other thing autocomplete, exactly. I just realized it's not really search in that you can't keyboard down to select a search result. But yes, I think we can remove this.

File: app/views/lists/index.html.erb [hunk 14]
```diff
@@ -4,36 +4,108 @@
...
+        matchingTags() {
+          return this.tags.filter(tag => tag.toLowerCase().includes(this.query.toLowerCase())).slice(0,10)
+        }
+      },
+      methods: {
+        showSelectedTag: function(listTags) {
+          return listTags.includes(this.selectedTag.toLowerCase()) || this.selectedTag === 'All Tags'
+        },
+        showQueriedTag: function(tags) {
+          return tags.includes(this.query.toLowerCase()) || this.query === ''
```
Comment (briankung) [inline app/views/lists/index.html.erb]: You're doing `.toLowerCase()` a lot. You could make `query` a computed property that lowercases the user's query. For the user's query, you could make it `rawQuery`. You could also make a `downcase` filter or method so you can reuse it on selectedTag (though I'm not quite sure where you're getting `selectedTag` from, yet)

Comment (briankung) [inline app/views/lists/index.html.erb]: Oh yeah, looks like we're getting rid of `selectedTag`

Comment (wsf246) [inline app/views/lists/index.html.erb]: selectedTag was for the select, which I'll get rid of

**Comment by briankung** [comment]
Using this is a little weird. Some notes:

1. Just by looking at it, I expect this to search title and description fields as well, not just tags. This is going to throw people off a lot imo.
1. It would be nice to be able to hit escape to clear
1. It might be nice to be able to hit down arrow to select an option - kind of like https://silviomoreto.github.io/bootstrap-select/examples/#live-search_1, though we want to start staying away from jQuery now
1. Right now, it takes an exact match on a tag to filter the lists. However, we could just filter the lists by whatever tags might match an incomplete query

It is nice how quickly it works, though
